apollo-runtime-kotlin/src/commonMain/kotlin/com/apollographql/apollo/executor/NetworkExecutor.kt

```diff
@@ -0,0 +1,90 @@
+package com.apollographql.apollo.executor
+
+import com.apollographql.apollo.ApolloError
+import com.apollographql.apollo.ApolloException
+import com.apollographql.apollo.context.DispatchersContext
+import com.apollographql.apollo.api.ApolloExperimental
+import com.apollographql.apollo.api.ExecutionContext
+import com.apollographql.apollo.api.Response
+import com.apollographql.apollo.network.NetworkRequest
+import com.apollographql.apollo.network.NetworkResponse
+import com.apollographql.apollo.network.NetworkTransport
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.catch
+import kotlinx.coroutines.flow.flatMapLatest
+import kotlinx.coroutines.flow.flow
+import kotlinx.coroutines.flow.flowOn
+import kotlinx.coroutines.flow.map
+
+@ApolloExperimental
+@ExperimentalCoroutinesApi
+class NetworkExecutor(
+    private val networkTransport: NetworkTransport
+) : RequestExecutor {
+
+  override fun <T> execute(request: ExecutionRequest<T>, executorChain: RequestExecutorChain): Flow<Response<T>> {
+    return flow { emit(request.toNetworkRequest()) }
+        .flatMapLatest { networkRequest -> networkRequest.execute(request.executionContext) }
+        .map { networkResponse -> networkResponse.parse(request) }
+        .flowOn(request.executionContext.ioDispatcher)
```

Comment:
Do we need the dispatcher to be configurable per-request ? I was thinking this could be a property of the `NetworkTransport` instead ?

Reply:
`ApolloClient` configures `RealApolloCall`:
```
class ApolloClient (...) {
  ...
  fun <D : Operation.Data, T, V : Operation.Variables> query(query: Query<D, T, V>): ApolloQueryCall<T> {
    return RealApolloCall(
        operation = query,
        scalarTypeAdapters = scalarTypeAdapters,
        executors = executors + NetworkExecutor(networkTransport),
        executionContext = executionContext
    )
  }
```

Then `RealApolloCall` configures `ExecutionRequest`:
```
val request = ExecutionRequest(
        operation = operation,
        scalarTypeAdapters = scalarTypeAdapters,
        executionContext = executionContext
    )
```

So dispatchers will be set only once in `ApolloClient` and will carry over to the request and every executor can get access to them. 

In this way we have one place where we can set dispatchers per `ApolloClient` instance.

Reply:
> So dispatchers will be set only once in ApolloClient and will carry over to the request and every executor can get access to them.

What other executor than the NetworkExecutor would require a dispatcher ? 

> In this way we have one place where we can set dispatchers per ApolloClient instance.

I was thinking that setting it on the `NetworkTransport` would make it more explicit to callers where the dispatcher is used. Obviously that doesn't work if other executors need to access it.

Reply:
After working on network layer I think we can drop passing any executor in the execution context for now. We can leave it to the client to decide, for example:
```
apolloClient
  .query(repositoriesQuery)
  .execute()
  .flowOn(Dispatchers.IO) // up to the client to provide this
  .single()
```

Reply:
Even better :+1: